Allow an optional vert at the beginning of a match branch

mdinger opened the pull request:
This was written to provide a formal proposal to allow a leading `|` at the beginning of a `match` as was discussed in https://github.com/rust-lang/rfcs/issues/1745. In that thread, it was previously stated the language team is [in favor of this change](https://github.com/rust-lang/rfcs/issues/1745#issuecomment-282150644).

Some of the sections I don't know anything about but I tried to fill in what I knew. For such a small topic, this form seems excessive and so a lot of the text seems fairly brief to me. Also, I'm not actually sure all the examples are needed but I thought it wouldn't hurt to add them.

In brief, this is a proposal that the following `match` syntax be legal:

```rust
enum E { A, B, C, D}
use E::*;

match foo {
| A
| B => (),
| C
| D => (),
}
```


---

[Rendered](https://github.com/mdinger/rfcs/blob/3b052a6311cc52d2767790576a8d8ebe443ccd53/text/0000-optional-match-vert.md)

joshtriplett commented:
@mdinger One nit: in example B, why not include the vert to make all the branches match, rather than calling attention to the lack of one with a comment?

More significantly, the "detailed design" section needs some details.  In that section, you should discuss modifying the grammar for match arms to allow an optional single vertical bar `|` before the pattern.

mdinger commented:
Example B was to show it is still legal to be inconsistent in that way but
even in that case, it doesn't make the logic unclear.

I'll looked for the grammar but didn't see it. I'll have to look again
another day.

On Feb 23, 2017 9:33 PM, "Josh Triplett" <notifications@github.com> wrote:

> @mdinger <https://github.com/mdinger> One nit: in example B, why not
> include the vert to make all the branches match, rather than calling
> attention to the lack of one with a comment?
>
> More significantly, the "detailed design" section needs some details. In
> that section, you should discuss modifying the grammar for match arms to
> allow an optional single vertical bar | before the pattern.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/1925#issuecomment-282186682>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AD9uOyfW3LuERRep_fXNmEYEqn5y6TjRks5rfkFggaJpZM4MKvkp>
> .
>

mdinger commented:
Actually, example B stemmed from a question on the original thread which I
think highlighted the possible confusion over something like that. So I
didn't think I should disregard it. Maybe it needs more comments

On Feb 23, 2017 9:43 PM, "Matthew Dinger" <mdinger.bugzilla@gmail.com>
wrote:

> Example B was to show it is still legal to be inconsistent in that way but
> even in that case, it doesn't make the logic unclear.
>
> I'll looked for the grammar but didn't see it. I'll have to look again
> another day.
>
> On Feb 23, 2017 9:33 PM, "Josh Triplett" <notifications@github.com> wrote:
>
>> @mdinger <https://github.com/mdinger> One nit: in example B, why not
>> include the vert to make all the branches match, rather than calling
>> attention to the lack of one with a comment?
>>
>> More significantly, the "detailed design" section needs some details. In
>> that section, you should discuss modifying the grammar for match arms to
>> allow an optional single vertical bar | before the pattern.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/rust-lang/rfcs/pull/1925#issuecomment-282186682>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AD9uOyfW3LuERRep_fXNmEYEqn5y6TjRks5rfkFggaJpZM4MKvkp>
>> .
>>
>

joshtriplett commented:
@mdinger In that case, you may want to show a "before" and "after" for example B: the inconsistent version, and the consistent version.

est31 commented:
I think for consistency it should also allow leading `,`, like it was suggested [in the issue](https://github.com/rust-lang/rfcs/issues/1745#issuecomment-272081745). Still, I think neither of the two notations should become official rust style.

keeperofdakeys commented:
I'm finding it a bit hard to read the examples when I glance over them. The indenting makes it see where each pattern starts or ends, and to spot where each match occurs.

So personally I don't think it would fit with the language unless it used a style like this:
```
match A {
    | A
    | B => println!("Give me A | B!"),

    | C
    | D => println!("Give me C | D!"),
}
```

nikomatsakis commented:
I very much appreciated the ability to have leading `|` in Ocaml (I've never used F#, though I've often wanted to) and I've also wondered why we don't support them in Rust from time to time, but never bothered to propose it. So I'm basically 👍 on this proposal.

That said, if we're going to **accept** leading `|` but not recommend you use them, it seems less exciting. I'd still be in favor, because macros and so on, but less so. Therefore, I'm also interested in the feedback of the "format" team -- I don't know best to cc them, though, so I'll just cc a few members that I can think of off the top of my head: @nrc, @solson, @joshtriplett. I'm not really following a lot of the discussions there, but last time I checked in, it seemed like we were aiming for a "less visual" more "block-oriented" formatting strategy. In that case, I feel like leading `|` might fit in there. (Also, I remember that rustfmt has some "issues" around match arms, maybe this helps to resolve them.)

joshtriplett commented:
@keeperofdakeys Allowing leading vert doesn't specify whether match arms should have a leading indent in addition to the `|` or not.

@nikomatsakis @rust-lang-nursery/style, because for some reason we put fmt-rfcs under rust-lang-nursery.

Given the availability of leading `|`, personally, I'd be inclined to recommend using it for any match pattern that contains an alternation.  For consistency, I'd personally suggest using it on the rest of the match patterns if any pattern in the match has an alternation in it.  (If every match arm consists of a single pattern, I don't think using it makes sense to prefix them all with `|`.)

mdinger commented:
To clarify, I never intended for this to be a proposal to change the recommended way matches are formatted. Stylistic topics are enormously difficult to push through to communities because everyone has their own preferences, which is okay and not problematic.

My intent was to showcase that allowing the leading vert could be a very valuable addition to some people without hindering understanding for others. If you don't like it, fine, don't use it but that is no reason to hinder others if they like it. The best I probably hoped for was for the vert to be allowed and an example in the docs somewhere include the style so that _people could decide if it is good_. That is, I don't always wait for rust-lang.org to tell me if I like a coding style. Sometimes, I decide for myself (it's nice to have a menu to show me what my options are though - I'd never have discovered this style if I hadn't seen it in F#).

@keeperofdakeys Styles are mutable. Your entire history in Rust and every other language colors your perception of what looks good. My history works the same way. I don't think I can categorically say that any simple style change of this sort could never look like Rust.

---

Taking an example from [regex](https://github.com/rust-lang/regex/blob/master/regex-syntax/src/parser.rs#L940-L957): can you identify which lines are match arms in less than a second in the following example?

```rust
            match c {
                '\\' => return match self.c() {
                            Some('#') => {self.advance(); Some('#')}
                            _ => Some('\\')
                        },
                '#'  => loop {
                            match self.c() {
                                Some(c) => {
                                    self.advance();
                                    if c == '\n' {
                                        break;
                                    }
                                },
                                None => return None
                            }
                        },
                _    => if !c.is_whitespace() {return Some(c);}
            }
```

Repeating the example, if I told you every arm started with a vert, could you do it faster? I can. I think they highlight them:

```rust
            match c {
                | '\\' => return match self.c() {
                            | Some('#') => {self.advance(); Some('#')}
                            | _ => Some('\\')
                        },
                | '#'  => loop {
                            match self.c() {
                                | Some(c) => {
                                    self.advance();
                                    if c == '\n' {
                                        break;
                                    }
                                },
                                | None => return None
                            }
                        },
                _    => if !c.is_whitespace() {return Some(c);}
            }
```

keeperofdakeys commented:
I actually think this would be a good addition to the language, I just wanted to point out that I found the RFC examples a bit hard to read. People are free to use a feature how they want, but I think that RFC examples are a good indication of how a feature could/should be used. Ultimately a style rfc is the way to decide how it should be styled though.

@mdinger I agree, it's easier to identify match arms in the second.

withoutboats commented:
When I have more variants than can fit on a line I write my matches like this:

```rust
match foo {
    Foo::Bar
    | Foo::Baz
    | Foo::Quux,
    => Whatever::One,
    Foo::Spam
    | Foo::Eggs,
    => Whatever::Two,
}
```

I would probably write this after this RFC:

```rust
match foo {
    | Foo::Bar
    | Foo::Baz
    | Foo::Quux,
    => Whatever::One,
    | Foo::Spam
    | Foo::Eggs,
    => Whatever::Two,
}
```
When I don't have a situation like this (which is most of the time), I make sure my matchrockets all align.

Most people write their matches differently from me, but I would still like to have the grammar be more flexible.

However, supporting `|` within patterns is more important, and so I want to make sure this is square with that proposal.

durka commented:
Unfortunately the closing braces in the same column kills it for me right away, along with the more minor issue @keeperofdakeys pointed out where it becomes less clear which lines are the beginnings of patterns and which are not.

mdinger commented:
@joshtriplett I added some updates to try to address your comments.
@keeperofdakeys I don't know how I can make the examples more clear. Branches always occur where the `=>` points. They always have and they always will. This appears to be finding the syntax unfamiliar and thus *feels hard* ([here's](https://upload.wikimedia.org/wikipedia/en/b/b1/Backwards_text_refdesk.png) something which technically should be really easy but probably will be really hard). The only example I think is somewhat difficult to parse is the first part of Example C. I find it tricky because the indentation is always changing somewhat how the line below keeps moving. I find the second part more consistent by far and much more straightforward.

![wavy](https://cloud.githubusercontent.com/assets/4156987/23336870/c0785aa6-fbaa-11e6-97d7-f17a3f75912c.png)

To others regarding braces, considering how syntax heavy rust is, I actually find the extra braces just a slight nitpick and isn't even that bad. Compare with the fairly common `)))))` which pops up at the completion of function calls. To the functional programmer, having tons of closing parenthesis is probably even more dumb than an out of place brace considering their languages don't even really require parenthesis. If Example C was translated to F#, most of the symbols (`,{}()::` just vaporize). Consider how clean this looks (not valid F#):

```rust
    match name
    | Name Franks name =>
        match name
        | Franks Alice
        | Franks Brenda
        | Franks Dave => FavoriteBook
            author: "alice berkley"
            title: "Name of a popular book"
            date: 1982
        | Franks Charles
        | Franks Steve => FavoriteBook
            author: "fred marko"
            title: "We'll use a different name here"
            date: 1960
    | Name Sawyer name  =>
        match name
        | Sawyer Tom => FavoriteBook
            author: "another name"
            title: "Again we change it"
            date: 1999
        | Sawyer Sid
        | Sawyer May => FavoriteBook
            author: "again another name"
            title: "here is a different title"
            date: 1972
```

---

I'm not saying all the extra symbols need to go away but when you compare it to something like this which, for the most part, should be somewhat clear (and less heavy-handed), the debate about whether 1 or 2 different braces should be aligned isn't that important. The language is syntax heavy and we do the best we can.

durka commented:
Um... I don't really like the argument "rust is already uglier than F#, so
it shouldn't matter if we make it a tiny bit more ugly here". I see the
advantages of this proposal, but I wonder if we can't find an indentation
style that doesn't make it harder to associate a given closing brace with
its opening.

On Sat, Feb 25, 2017 at 11:18 PM, mdinger <notifications@github.com> wrote:

> @joshtriplett <https://github.com/joshtriplett> I added some updates to
> try to address your comments.
> @keeperofdakeys <https://github.com/keeperofdakeys> I don't know how I
> can make the examples more clear. Branches always occur where the =>
> points. They always have and they always will. This appears to be finding
> the syntax unfamiliar and thus *feels hard* (here's
> <https://upload.wikimedia.org/wikipedia/en/b/b1/Backwards_text_refdesk.png>
> something which technically should be really easy but probably will be
> really hard). The only example I think is somewhat difficult to parse is
> the first part of Example C. I find it tricky because the indentation is
> always changing somewhat how the line below keeps moving. I find the second
> part more consistent by far and much more straightforward.
>
> [image: wavy]
> <https://cloud.githubusercontent.com/assets/4156987/23336870/c0785aa6-fbaa-11e6-97d7-f17a3f75912c.png>
>
> To others regarding braces, considering how syntax heavy rust is, I
> actually find the extra braces just a slight nitpick and isn't even that
> bad. Compare with the fairly common ))))) which pops up at the completion
> of function calls. To the functional programmer, having tons of closing
> parenthesis is probably even more dumb than an out of place brace
> considering their languages don't even really require parenthesis. If
> Example C was translated to F#, most of the symbols (,{}():: just
> vaporize). Consider how clean this looks (not valid F#):
>
>     match name
>     | Name Franks name =>
>         match name
>         | Franks Alice
>         | Franks Brenda
>         | Franks Dave => FavoriteBook
>             author: "alice berkley"
>             title: "Name of a popular book"
>             date: 1982
>         | Franks Charles
>         | Franks Steve => FavoriteBook
>             author: "fred marko"
>             title: "We'll use a different name here"
>             date: 1960
>     | Name Sawyer name  =>
>         match name
>         | Sawyer Tom => FavoriteBook
>             author: "another name"
>             title: "Again we change it"
>             date: 1999
>         | Sawyer Sid
>         | Sawyer May => FavoriteBook
>             author: "again another name"
>             title: "here is a different title"
>             date: 1972
>
> ------------------------------
>
> I'm not saying all the extra symbols need to go away but when you compare
> it to something like this which, for the most part, should be somewhat
> clear (and less heavy-handed), the debate about whether 1 or 2 different
> braces should be aligned isn't that important. The language is syntax heavy
> and we do the best we can.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/1925#issuecomment-282531877>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAC3n6LhrOgW9LRSgllKDq2B9j6t1RHVks5rgP0fgaJpZM4MKvkp>
> .
>

mdinger commented:
Okay. I think I've done my part trying to defend the stylistic part of the proposal and people who don't like it still won't like it. Having said that, I don't know where the consensus stands as far as the technical part of this being supported by the compiler. I haven't seen any comment which states making this legal is problematic in any fashion aside from something stylistic.

Let me know if I need to update the RFC in any other ways.

dashed commented:
If this gets approved, I hope there will be an associated rustfmt RFC to helps folks that may either approve or disapprove this style: https://github.com/rust-lang-nursery/fmt-rfcs

nrc commented:
I don't think we would recommend them - we don't recommend a leading comma style, for example. I know @joshtriplett favours leading newlines with punctuation, e.g.,
```
a + b
+ c + d

// c.f.

a + b +
c + d
```

But that is not what is being proposed here. I can't think of anywhere we are proposing actually leading with punctuation.

In general, I would be in favour of this RFC - I usually prefer more flexibility in our syntax rather than less, but I don't think we would make this the default style. In particular, the needing a `|` in the single pattern case seems very weird to me.

(Do we allow trailing `|`? That seems useful for macros, but I don't think we'd recommend a trailing `|` either).

joshtriplett commented:
@mdinger For the sake of demonstrating the smallest incremental change, I'd suggest changing the examples so they still indent match arms as normal (four spaces in front), rather than outdenting them and treating the vert-and-space as their indent.  We could evaluate the latter as a potential style change, but that leads to the issue with trailing braces.  Let's not conflate that style change and associated brace indentation issue with the minimal grammar proposal.

You could add a section towards the end mentioning that potential style change, move the discussion of the brace problem from the drawbacks to that new section, and punt the actual style, arguing that the grammar change is useful independently.

mdinger commented:
I don't mind doing that. I might get around to it a day or so but if not then, it might be a few weeks. Might be pretty busy and don't know what my free time will be like.

nikomatsakis commented:
I think the reason to make this a recommended style would be to avoid the double indentation and so forth. But I agree it's a separate thing-- still it seems like it'd be good to take this idea into account when settling on the best indentation style for matches

vitiral commented:
In elm you would do something like

```
match Z
    { A
    | B =>... 
    , C
    | D
   ... 
} 
```

Which would be legal rust syntax and make everything align.

joshtriplett commented:
In the interests of moving this forward:

I'm not seeing any objections to the idea of supporting this in the grammar. I do see some comments about what our recommended style should look like, and in particular whether or not this should replace indentation of match arms.  The thing that's causing those two to be conflated is that the RFC's samples all assume a particular style when writing, and not everyone is a fan of that style.  As far as I can tell, the objections to this all boil down to "the style as shown in the RFC doesn't look appealing", which I wouldn't argue with.

Does anyone object to allowing the optional vert *at all*?  If not, I would propose that we move to accept this, and then separately determine how it might affect the default style.

(And with my style team hat on, I'd be inclined to *not* change the recommended indentation, but instead just add the starting vert for alignment purposes when writing a series of patterns one per line.)

mdinger commented:
I apologize for the delay. I went out of town and haven't looked at this in weeks. I just pushed a flurry of patches pushing this towards grammar improvement and avoiding the style discussion entirely. I additionally added the trailing vert as there is benefit to doing so and little if any detriment to doing so as far as I have seen from this discussion.

I added a macro section and I was hoping someone could supply me with a simple example which would benefit from preceding and/or trailing verts. If not, I can remove it but it seemed useful to include a pertinent example in the RFC.

@vitiral That is clever though kinda quirky.

mdinger commented:
@joshtriplett In case you missed my last comment, this could still use another review.

joshtriplett commented:
@mdinger Just reviewed the updated version.  Looks great to me.

(I don't think I'd ever want to use *both* the leading and trailing vert, but both independently seem reasonable to me.)

mdinger commented:
Thanks for reviewing.

If anyone has a macro example which would benefit from this RFC (people suggested macros would benefit), please include it in a comment because it seems useful to include in this RFC but I don't have an example.

nikomatsakis commented:
@rfcbot fcp merge

OK, I think there isn't much controversy here, so I'm going to move to **accept** this RFC. In summary, the proposal is to permit a leading`|` in patterns, primarily because this will make macro generation and so forth easier (every pattern can be emitted as `| pat`). This may or may not affect the choices that the rustfmt team makes (likely not, based on the comments here).

withoutboats commented:
There are a lot of thumbs down on the opening comment, but little expressed negative feedback.I wonder if anyone wants to  give feedback explaining why they think this shouldn't be accepted?

I think its clear that this isn't going to be the recommended style, but in general we've taken a very liberal approach to the grammar, accepting all kinds of things that would be bad style (as Niko says the most practical reason for this is to make writing macros easier).

steveklabnik commented:
I'm not aware of this being a particular difficulty for macros, and I strongly prefer a standard style, so adding a way to write non-standard code without solving some kind of codegen pain means I voted 👎

mgattozzi commented:
I don't really like the style, nor am I aware of any macros that this would help. To merge this RFC as is without an example as to where it helps in macros or providing any kind of better justification I think is a mistake. All this RFC seems to do is change the style in a way that serves to scratch a few peoples' itch about something so that things line up. It just seems odd to have something like

```rust
match foo {
    A |
    B | => (), // At first glance it seems like a syntax error. 
               // Where's the other type for after this |? | implies this pattern or this one match
               // and it's weird to just have => (), come after it
    C |
    D | => (),
}
```

be legal all for the sake of style. I'm against this personally. If the team does decide to accept it then I feel like the RFC as it stands could be a bit more polished particularly the detailed design part with some mentorship as it seems unsure as to how to implement it. It should also, as I have mentioned before, have examples of the Macros example it claims it would help.

joshtriplett commented:
@mgattozzi I'm not a fan of the "trailing vert" approach.  Leading vert, though, does help:

```rust
match foo {
    | A
    | B
    | C => {
        ...
    }
    | D 
    | E => {
        ...
    }
    | F
    | G => value,
}

withoutboats commented:
@steveklabnik @mgattozzi Your argument makes sense, but I feel that its philosophically misaligned with the way we normally make decisions about the grammar & what I expect to behave.

For example, even though it repeats bounds, contains empty bounds, repeated boiunds, unnecessary parens, trailing `+`, unused HRTBs, etc this function is grammatically correct & even well typed:

```rust
fn foo<T: , U: ToString + ,>()
where
    T:,
    T: ToString + ,
    T:,
    (U): ToString +,
    T: for<'a,> ToString +,
    U:
{ }
```

I myself would definitely make use of leading `|` (stylistically recommended or not) not because it 'scratches an itch' but for the same reason I love trailing commas. When I have a 'collating' match, where a bunch of variants are pulled together into a smaller number of arms, it produces cleaner diffs and easier editing (at least with a line-oriented editor) to structure it like this:

```rust
match foo {
    | Foo::Bar
    | Foo::Baz
    | Foo::Quux,
    => Whatever::One,
    | Foo::Spam
    | Foo::Eggs,
    => Whatever::Two,
}
```

It doesn't matter which variant is first if they all begin `|`. Obviously other people have different opinions on this, just like people have different opinions on trailing commas, but I think rustc should be unopinionated on issues like this and so far that's the position we've tended to take.

mgattozzi commented:
That's fine and I totally understand where you're coming from with this. Like you said, we just disagree. I do think you're right in Rust being opinionated about certain things but in this case I just felt the syntax change didn't gain much from being changed like this, especially if we have a tool like rustfmt which is trying to have a consistent style. It just seems a bit at odds in some sense. Though I'm just one person so if a lot more people prefer it (or the core team gives it a go) then I say it should be merged.

nikomatsakis commented:
I have personally found it annoying that I can only generate `|` "in between" patterns when writing macros. (Do we support a trailing `|`?)

pnkfelix commented:
@rfcbot concern trailing-vert

I think a number of people do not like supporting trailing vert. I hit the FCP check-box originally thinking that this RFC was solely about adding support for leading vert (because that was the draft I had originally read), but then I double-checked and realized that the RFC author had added trailing vert to the spec as well.

aturon commented:
@pnkfelix Oh dear, I also missed that change, and am also concerned about it (and it appears others on thread are as well).

ubsan commented:
I definitely want leading vert, and very much do not want trailing vert.

durka commented:
Okay, I'll bite... I haven't seen anyone explain *why* trailing vert is
worse than leading vert.

On Fri, Apr 28, 2017 at 4:05 PM, Nicole Mazzuca <notifications@github.com>
wrote:

> I definitely want leading vert, and very much do not want trailing vert.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rust-lang/rfcs/pull/1925#issuecomment-298094176>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAC3n9-y0PHLx2QTfp1kPFJyZrt9n5O2ks5r0kabgaJpZM4MKvkp>
> .
>

ubsan commented:
@durka style.

aturon commented:
@durka There are a few arguments. The most compelling to me is thinking by analogy to e.g. trailing comma, where you get consistency *on each line*:

```rust
StructName {
    field1,
    field2,
    field3,
}
```

Since we usually write `|` at the *beginning* of the next line, the analogous thing is to have an extra one at the *beginning* of the first line:

```rust
match expr {
    | Pattern1 => { .. }
    | Pattern2 => { .. }
    | Pattern3 => { .. }
}
```

whereas it doesn't make much sense at the end:

```rust
match expr {
    Pattern1 = > { .. }
    | Pattern2 = > { .. }
    | Pattern3 => { .. }
    |
}
```

And of course, to the extent that this is just about macros, either way solves the problem.

solson commented:
@aturon Trailing vert is not my preference, but I think this is more accurate as the ideal style for it:

```rust
match expr {
    Pattern1 |
    Pattern2 |
    Pattern3 |
    => { .. }
}
```

Which has the same consistency benefits.

aturon commented:
@solson Understood. What I was trying to get at is, given *today's* style for `|` (which I think is very unlikely to change), there's a simple reason to prefer leading, and little reason to allow for both.

solson commented:
Ah, hm, I didn't realize that was the argument. For example, the following looks very normal to me:

```rust
match expr {
    Pattern1 |
    Pattern2 |
    Pattern3 => { .. }
}
```

I guess I'm out of the loop about what "today's style" is.

le-jzr commented:
I see the discussion as conflating two distinct problems.

- What should the grammar allow.
- What style should be preferred / enforced.

Personally, I would consider the possibility of both trailing and leading `|` to be a boon for macros and code generation, regardless of the intended style of handwritten code. Besides, aesthetics are hard to evaluate without experience with either variant. Unless there is a compelling reason not to allow it on technical grounds, stylistic discussions are of limited value, as long as there is a tool like `cargo fmt` that formalizes the "one true coding style".

withoutboats commented:
I tend to agree with @le-jzr. Similarly I wouldn't mind making leading commas grammatic, as abhorrent as I personally find styles like this:

```rust
    struct Foo {
    ,   bar: 0
    ,   baz: 0
    }
```

But I'm also not going to go to bat for either of these so if other people feel strongly that we should only allow trailing commas and leading verts, that's fine with me.

mdinger commented:
FWIW, when I read people having concern about the trailing vert, I read it as if something really bad could happen if the vert was supported. What is the terribly awful thing that could happen: it is possible people could (no might) use bad style 👻 👻 👻 (spooky)

Also, I don't think I'm a fan of the trailing vert but I also don't see where the dislike of it as a feature comes from. I added it under the understanding macros could benefit from it, it really doesn't hurt anyone else, and some people may like it in other cases as well.

aturon commented:
@mdinger Adding leading vert already addresses the macro use case, no? Similarly, we don't permit leading commas.

While permitting "bad" styles isn't such a big deal, there's also little reason to go out of our way to make room for more styles in the grammar, especially given that we're hoping rustfmt will help us, as a community, *consolidate* our styles.

mdinger commented:
I just wanted to mention it because this whole thread reads quite fearfully about such things. In regards to the macro case, I assume so because of the conversation up to this point but as I am not as familiar with macros as others and don't know of any specific cases. I had hoped someone would put forth a specific macro example showcasing the benefits but none have surfaced yet.

le-jzr commented:
If we want this merely as a convenience for code generation, and not addition of style options, then we should go with trailing vert and no leading vert, since it's​ consistent with comma. The only argument for using leading vert over trailing is "I like the way it looks in code", and that's an arbitrary and subjective choice for something we probably don't even want to see in handwritten code.

withoutboats commented:
I do intend to use leading vert in handwritten code.  :)

le-jzr commented:
I rescind my argument. :)

le-jzr commented:
But despite my what I wrote earlier, I do think adding both is entirely unnecessary, and it's always prudent to add only as little as needed. After all, we can add the other case later, but we can't remove it once it's there. There is always gonna be some disagreement -- some want to use leading, some want to use trailing, and some would be unhappy if it was up to choice.

nikomatsakis commented:
@rfcbot fcp cancel

Given the concerns about leading/trailing bars, I'm going to cancel my request to merge. I'm not really sure what the consensus is now so I think better to restart.

joshtriplett commented:
As far as I can tell, there's a consensus supporting leading vert, and I think the intent was to accept an RFC for that. That got derailed by a suggestion to add trailing vert as well, which doesn't have consensus.

Can the author please submit a version of the RFC that only includes leading vert? I think we can quickly FCP and accept that one.

If you feel strongly about trailing vert, or someone else does, let's deal with that in a *separate* issue, please. I don't think that'll get consensus as easily, though.

mdinger commented:
@joshtriplett I have updated the RFC as requested and removed the inclusion of the trailing vert. The majority of the RFC remains similar otherwise. Thanks for keeping up with this.

joshtriplett commented:
Given that, prior to the concerns about trailing vert, we were at FCP with almost everyone on the lang team agreeing to it, can we go back into FCP directly?

petrochenkov commented on text/0000-optional-match-vert.md:
```diff
@@ -0,0 +1,257 @@
+[design]: #detailed-design
+
+I don't know about the implementation but the grammar could be updated so that
+an optional `|` is allowed at the beginning and the end. Nothing else in the
```
The trailing `|` is still mentioned here.

mdinger replied:
Fixed

petrochenkov commented:
Together with https://github.com/rust-lang/rfcs/pull/1882 it will allow some fun code like
```
let | a = 10;
```
, nothing new though, inappropriate use of trailing separators always look weird.

mzji commented on text/0000-optional-match-vert.md:
```diff
@@ -0,0 +1,257 @@
+- Feature Name: `match_vert_prefix
```
Looks like the trailing backquote charactor is lost.

mdinger replied:
Fixed

aturon commented:
@rfcbot fcp merge

joshtriplett commented:
(Sorry, wrong label.)

liigo commented:
Weird. (The optional vert at the beginning is not very useful. I double people often writing `X|Y|Z => { }` many times. In most situations each enum variant need its own match branch, that's why they're different variants. And you can always write `X|Y|Z` in the same line.)

```rust
enum E { X, Y, Z}
use E::*;

match foo {
| X => (),
| Y => (),
| Z => (),
}
```

ubsan commented:
@liigo people downvoting it didn't like trailing vert, afaict. Also, you're ignoring:

```
match foo {
  | X
  | Y
  | Z
  => {
    println!("hi");
  }
```

main-- commented:
>That said, if we're going to accept leading `|` but not recommend you use them, it seems less exciting.

This is exactly what I'm worried about.

* When I first encountered this style in F# it was very confusing to me - why begin every match arm with `|` when it's only single alternatives?
* I also believe that it doesn't really pull its weight: Admittedly, I don't write an awful lot of Rust code but after several years, I can't remember a single match pattern with too many alternatives for a single line - which is, as I understand, the only case that directly benefits from leading verts. Correct me if I'm wrong, but isn't the entire reasoning that because it's more consistent *in this particular case*, we would extend that to *every match everywhere*? Seems questionable if only, like < 10^-3 of all match arms are improved by this change.

So to summarize, I don't mind leading `|`, but I do mind a style recommendation that uses it everywhere.

joshtriplett commented:
@main-- Now that it's available, we'll be evaluating it within the style team to determine whether and where to recommend it.

Personally, I have *no* intention of recommending its use for single patterns, only for multiple patterns split across lines.

kennytm commented:
Just to clarify, I downvoted against *both* the leading and the trailing `|`. @ubsan

liigo commented:
@ubsan 

when matching `X|Y|Z` along side `A`, `B`, `C`, it still looks wired (either use leading `|` or not).

```
match foo {
  | A => {}

  | B
  => {}

  C => {}
  D => {}

  | X
  | Y
  | Z
  => {
    println!("hi");
  }
}
```

leonardo-m commented:
An alternative (and possibly better?) solution is to not mud Rust syntax and give Rust macros the ability to solve this problem.

A nice quote from John Carmack:

> Everything that is syntactically legal that the compiler will accept will eventually wind up in your codebase.

liigo commented:
> The RFC got 9 :+1:  and 14 :-1:. The result should have been considered by the team.
> **update**: now the vote result is: 12 :+1: and 20 :-1: .

**update 2** (2017-08-21): 15 :+1: VS 28 :-1:.

aturon commented:
This RFC has been **merged**! Thanks @mdinger for writing it, and everyone else for the discussion.

[Tracking issue](https://github.com/rust-lang/rust/issues/44101)

mdinger commented:
@joshtriplett Thanks for seeing that this got followed through with. I don't know how you guys keep your head in it with some of these RFCs. They can be endlessly frustrating. I'd have given up.

est31 commented:
@mdinger could you have a look and consider signing off: https://github.com/rust-lang/rfcs/issues/2078

liigo commented:
![updown](https://user-images.githubusercontent.com/346530/29756241-0374d734-8bd5-11e7-8ced-e140c76cf010.png)

It's too sad that, many people dislike this one (-31 VS +15), but it was still approved.
To the core team, are you still keep open? are you still listening to the community?

ubsan commented:
@liigo if people do not make their voice heard, we can't help that. It has strong support across the people who care about this stuff.  The thumbs up/thumbs down thing is not a voting system, and expecting us to treat it that way seems unreasonable. The only objection I can see is "it looks weird", which is completely subjective? Otherwise, testing it out seems like a good thing. If people don't like it, we can remove it.

leonardo-m commented:
> The thumbs up/thumbs down thing is not a voting system, and expecting us to treat it that way seems unreasonable.

Then please remove the voting system now, don't give illusions.

eddyb commented:
@leonardo-m How? GitHub, AFAIK, doesn't allow us to remove reactions.

leonardo-m commented:
> How? GitHub, AFAIK, doesn't allow us to remove reactions.

Ignore my comment then, and sorry.

est31 commented:
@eddyb github does allow that. Just click at the reaction symbol.

eddyb commented:
@est31 I meant to disable them. Not someone removing their reaction.

est31 commented:
@eddyb ah right, that seems impossible indeed.

withoutboats commented:
We definitely don't make decisions on the basis of thumb counts. Obviously they can have an impact, but we try to make decisions in a more nuanced manner. Here are reasons that I personally think the RFC should be merged:

* Despite the thumb count, the comments felt much more positive. What seemed clear is that while many users didn't want this syntax (and maybe wouldn't allow it in their code base), there were some users who very strongly did want this and were willing to press on it. I weighed these strong positive feelings more heavily than the mild negative feelings.

* We tend to support a flexible grammar when possible, including things we don't consider idiomatic at all. Philosophically, this is an RFC we are inclined to accept, and there'd probably have to be a pretty strong, practical reason for us not to accept it.

* Sort of along the lines of both of the previous points, I have difficulty imagining this having a serious negative impact on any user. You might see some match arms you don't like, maybe find them a little confusing for a moment, but I have difficulty valuing this impact over the positive impact on people who want to write their code this way.

Speaking more general, RFCs are a deliberative process, not an up or down vote. The point of the process is to seek out useful commentary from community members, and while thumbs provide some feedback for decision making, they're limited in their impact in proportion to the amount of information they contain.

steveklabnik commented:
@liigo Small procedural note:

> To the core team, are you still keep open?

This is not a core team RFC, but a lang team one.

jansegre commented:
I realize I'm a bit late, but I just saw this on TWiR. I haven't read the following point: from what I understand one of the barriers for people to get into Rust is the syntax, and those people would probably be even more put off if they encounter this kind of style disagreement upfront (this syntax is being approved because people want to use it, and when it is stable it may end up on popular posts and even on official docs) and since they are not into Rust yet, most won't participate in the RFC process.

__To summarize that, I think this has some (potentially) negative impact on people outside of Rust who don't participate on the RFC process.__

Personally I just find it a bit ugly and confusing, my instinct is to expect something before a `|` so it looks like a condition is missing. I also think that reasonable style (placing `|` on the end of the previous line) was already proposed that makes the code align. I get the flexibility argument, for example, one could use leading semicolons, for example, but I also feel that this is a stretch.

lilianmoraru commented:
I just saw this in TWiR - this is a really weird syntax.....

chris-morgan commented:
(Somehow I missed this one in the last couple of weeks of TWiR or I would have commented earlier. I don’t really get the motivation for this and I feel the drawbacks section being empty is unreasonable, but let those pass for now.)

One concern I have about this is the impact on macro rules. As it stands, we have the `pat` fragment specifier which dovetails well with repeating groups to cover *most* of `match_pat` in a very simple pattern—​granted, it doesn’t support the guard (`[ "if" expr ] ?`), but at least it supports multiple patterns (`pat [ '|' pat ] *`):

    $($patterns:pat)|+

If you want to support the `[ "if" expr ] ?` as well, it’s not *too* bad: you need to tack on

    $(if $guard:expr)*

(Side note: we really need `$(…)?`.)

After this change, people writing macros that want to accept match patterns will need to prefix `$(|)*` to make it support the full `match_pat` grammar. It’s not a *big* deal, but it’s just one more thing that stands in the way of writing DSLs that behave like proper Rust.

You may say that trailing commas already have this problem (hence the pattern `$(x:expr),* $(,)*`, the `$(,)*` part of which I freely admit many don’t know about—so it *is* already a problem), but I think it’s more severe (though less common) with this one because I don’t expect many people to write match patterns with leading pipes (I never intend to), and so people writing macro libraries are much less likely to think of coping with the optional leading pipe, which will make people that like this pattern grouchy.

One of the main problems with increasing grammar complexity is in macros that wish to support or emulate the full grammar with certain variations of behaviour; unless we have a fragment specifier which handles that part of the grammar, macro rules become steadily more complex, or become incomplete (leading to frustration when the macro doesn’t work with otherwise normal Rust code).

After this RFC is implemented, full support for match patterns in a macro are going to be

    $(|)* $($patterns:pat)|+ $(if $guard:expr)*

juggle-tux commented:
i really dislike the idea to have many syntax options for the same think. It makes it imo harder for a beginner to understand the language since it raises questions like
“With syntax to use and why?”, “Are there any differences and if not why have the different syntax in the first place?”, “With way will run faster?” and so on

SoniEx2 commented:
Ppl don't use 

```rust
match x {
    Hello     |
    World     |
    AndThings => ...,
}
```

?

You know we have this same "issue" with `||` and `&&` and so on, right?

*Wonders how long until Rust decides to deprecate this*

The **only** nice thing about this is that now you can "convert" C into Rust with a simple `s/case [ident]:/| [ident]/` but if that's the only nice thing and everything else sucks then it's not really a good feature now is it?

SoniEx2 commented:
```rust
if
x
|| y
|| z
{
```

Are we gonna allow this next, for consistency?

```rust
if
|| x
|| y
|| z
{
```

I still prefer the apparently much dreaded

```rust
if
x ||
y ||
z {
```

I have absolutely no idea why this is so dreaded.

SoniEx2 commented:
I'll take the clippy lint tbh

H2CO3 commented:
@warlord500 indeed, inside an expression, leading `||` means the beginning of a closure with no arguments. So if that's allowed, Rust (and we) will have a bad (as in C++ context-sensitivity level bad) grammar problem.

est31 commented:
The feature was stabilized and shipped in the 1.25.0 version of the Rust compiler/language. You won't be able to remove it now due to the stability promise, except of removing it from the 2018 edition but I'd say that is very unlikely... It'd require a separate RFC and good reasoning (not just thumbs). So fuming is a bit pointless I'd say.

mgeisler commented:
@est31 

> So fuming is a bit pointless I'd say.

I think the fuming should be seen as a signal to the language team that they might want to be more careful about adding such features in the future.

Perhaps the down votes should be seen as a "Hey, let's stop and reevaluate this — out of the people who voted, a vast majority didn't like this idea". A good next step would be to wait until someone presents proper arguments against the idea. Accepting an idea by default seems bad to me.

yarrow commented:
@mgeisler 
> Perhaps the down votes should be seen as a "Hey, let's stop and reevaluate this — out of the people who voted, a vast majority didn't like this idea".

That's what @withoutboats did in [a comment last year](https://github.com/rust-lang/rfcs/pull/1925#issuecomment-293681943) after the first call for FCP:

> There are a lot of thumbs down on the opening comment, but little expressed negative feedback. I wonder if anyone wants to give feedback explaining why they think this shouldn't be accepted?

That resulted in re-opening the discussion. Posts between the first and the second FCP mostly addressed concerns about trailing vertical bars. Trailing vertical bars were removed from the RFC.

In the second FCP itself, two people said they were against it. One gave no reason.  The second gave two: that it was weird syntax, and that there were more thumbs down than thumbs up.   The thumbs up/down is not a voting mechanism, and in this case it was also hard to know whether the thumbs down were primarily about leading or trailing vertical bars.

As for the growing number of thumbs down, I can't know whether a new thumbs-down means "I'm against leading vertical bars" or "I'm annoyed with the language team". And neither can the language team. A comment conveys a lot more information than a thumbs-down.

I think the language team did a fine job of listening to the community on this PR. It may be that
* it was the wrong decision;
* Rust should make syntax changes much more slowly than it has in the past; or
* the process for involving the community could be improved.

I don't know the answer to the first two points.  The answer to the third point is of course "Yes, the process could be improved." Because anything can be improved.  And I do think the process is currently excellent.  (I also believe that treating thumbs up/down as a vote would be a move in the wrong direction: online polls are not a good way to count votes, let alone to reach a rough consensus.)

est31 commented:
@mgeisler like any kind of feedback, complaining and thumbs down have the most effect the earlier on you are in the process. After a feature has been stabilized and even shortly before it is being stabilized, complaints have the least effect.

You didn't even define what "such features" means... how should the language team know that?

What I suggest for you in the future is that you read the RFCs opened and voice your concerns, or even just :-1: the RFCs you don't like. The `try` RFC #2388 is on a similar course as this one: the number of :-1: is increasing more rapidly than the number of :+1: 's. But better is to :-1: earlier on in the process.

SoniEx2 commented:
`try` / #2388 can still be pulled back, at least, since it's nowhere near stable

BurntSushi commented:
I think this thread has outlived its usefulness.
